### PR TITLE
gh-85858: Remove PyUnicode_InternImmortal() function

### PR DESCRIPTION
**Pull request by vstinner**
Remove the PyUnicode_InternImmortal() function and the
SSTATE_INTERNED_IMMORTAL macro.

The PyUnicode_InternImmortal() function is still exported in the
stable ABI. The function is removed from the API.

PyASCIIObject.state.interned size is now a single bit, rather than 2
bits.

Keep SSTATE_NOT_INTERNED and SSTATE_INTERNED_MORTAL macros for
backward compatibility, but no longer use them internally since the
interned member is now a single bit and so can only have two values
(interned or not interned).

### CODE REVIEW

**Comment by vstinner** [comment]
@methane @serhiy-storchaka: Would you mind to review this change?

I reused https://github.com/python/cpython/issues/85858 issue.

File: Include/cpython/unicodeobject.h [hunk 1]
```diff
@@ -147,15 +147,9 @@ typedef struct {
...
     struct {
-        /*
-           SSTATE_NOT_INTERNED (0)
-           SSTATE_INTERNED_MORTAL (1)
-           SSTATE_INTERNED_IMMORTAL (2)
-
-           If interned != SSTATE_NOT_INTERNED, the two references from the
-           dictionary to this object are *not* counted in ob_refcnt.
-         */
-        unsigned int interned:2;
```
Comment (serhiy-storchaka) [inline Include/cpython/unicodeobject.h:109]: I would keep it 2-bit for a time.

File: Include/unicodeobject.h [hunk 2]
```diff
@@ -265,10 +265,6 @@ PyAPI_FUNC(PyObject *) PyUnicode_InternFromString(
     const char *u              /* UTF-8 encoded string */
     );
 
-// PyUnicode_InternImmortal() is deprecated since Python 3.10
-// and will be removed in Python 3.12. Use PyUnicode_InternInPlace() instead.
-Py_DEPRECATED(3.10) PyAPI_FUNC(void) PyUnicode_InternImmortal(PyObject **);
```
Comment (serhiy-storchaka) [inline Include/unicodeobject.h:261]: Is not 2 versions too short period for removing from the public C API? I think we should be more conservative in such cases and prolong the deprecation to at least 4 years.

I would keep it as an alias of `PyUnicode_InternInPlace` which may print a deprecation warning.

Comment (methane) [reply Include/unicodeobject.h:261]: There are no use from top4000 packages.
https://github.com/hpyproject/top4000-pypi-packages/search?q=PyUnicode_InternImmortal

We can remove it in early alpha and resurrect it if some packages are broken by this.

**Review by methane — APPROVED** [review]

File: Include/cpython/unicodeobject.h (hunk 1 above)
Comment (vstinner) [reply Include/cpython/unicodeobject.h:109]: PyASCIIObject is not part of the stable ABI. Why do you care about keeping 2 bits for interned?

File: Include/unicodeobject.h (hunk 2 above)
Comment (vstinner) [reply Include/unicodeobject.h:261]: > Is not 2 versions too short period for removing from the public C API?

I followed PEP 387 process.

> I think we should be more conservative in such cases and prolong the deprecation to at least 4 years.

Why do you care of this function in specific? Are you aware of projects using it?

> There are no use from top4000 packages.

Same when I looked in December 2021:
https://github.com/python/cpython/issues/85858#issuecomment-1093882377

History of this function: https://github.com/python/cpython/issues/85858#issuecomment-1093882364

> We can remove it in early alpha and resurrect it if some packages are broken by this.

Sure. I did that multiple times and I'm fine with reverting if it causes too many issues.

Removing the function early in 3.12 dev cycle gives more time to users to detect if their project is affected, to give them more time to decide how to update their code, or to ask to revert the change.

**Comment by vstinner** [comment]
I rebased my PR to fix merge conflicts.

**Comment by vstinner** [comment]
I merged my PR. Thanks for your review @methane.

@serhiy-storchaka: I understand your concerns. I tried to reply to them. I rely on @methane's approval to merge my change ;-)

**Comment by SnoopJeDi** [comment]
Do I understand correctly that this changeset alters the alignment of `PyASCIIObject` and derived types? I was just looking around this part of CPython and noticed that the embedded struct has a size of 31 bits (= `1 + 3 + 1 + 1 + 25`), where it used to be 32 bits (= `2 + 3 + 1 + 1 + 25`).

I'm not at all familiar with the m68k issue that caused the addition of the padding field (nor what the implications might be on target platforms), but thought that I would ask for confirmation if the adjustment to the width of `interned` here without a corresponding adjustment to the padding was deliberate.

**Comment by methane** [comment]
@SnoopJeDi Nice catch.
FYI, 31bit is enough to guarantee 32bit alignment, so this nit doesn't cause any real bug.

**Pull request merged**